module/mailnews/mailread.c

```diff
@@ -353,3 +395,153 @@ void UserMailReply(int msg_num, Bool reply_all)
    MakeReplySubject(reply->subject, MAX_SUBJECT);
    MailSendReply(hReadMailDlg, reply);
 }
+
+void OnColumnClick(LPNMLISTVIEW pLVInfo)
+{
+   int nSortColumn = pLVInfo->iSubItem;
+   bool bSortAscending = true;
+
+   if (nSortColumn == currentSortColumn)
+   {
+      // Toggle sort direction if the same column is clicked again
+      currentSortAscending = !currentSortAscending;
+   }
+   else
+   {
+      // Set ascending order for new column
+      currentSortColumn = nSortColumn;
+      currentSortAscending = true;
+   }
+
+   LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+   if (hbmUpArrow != NULL && hbmDownArrow != NULL)
+   {
+      ListView_SetHeaderSortImage(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+   }
+
+   // Sort list
+   ListView_SortItemsEx(pLVInfo->hdr.hwndFrom, CompareMailListItems, lParamSort);
+}
+
+/****************************************************************************/
+/*
+ * CompareMailListItems: Comparison function for sorting items in the dialog
+ * list view.
+ */
+int CALLBACK CompareMailListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   bool bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+   char szBuf1[MAX_SUBJECT], szBuf2[MAX_SUBJECT];
+   int time1 = 0;
+   int time2 = 0;
+
+   HWND hListView = GetDlgItem(hReadMailDlg, IDC_MAILLIST);
+
+   // Extract the msg_num values from the list view items
+   ListView_GetItemText(hListView, lParam1, COL_ORDER, szBuf1, sizeof(szBuf1));
```

Comment:
> Note: During the sorting process, the list-view contents are unstable. If the callback function sends any messages to the list-view control aside from LVM_GETITEM (ListView_GetItem), the results are unpredictable.

Noting this here in case it raises any red flags. It would seem to me like rescanning would be the only existing functionality that _could_ interfere, but doesn't.